### PR TITLE
config_tools: add CPU capability and BIOS invalid setting checks

### PR DESCRIPTION
**Pull request by Kunhui-Li**
1. Add BIOS invalid setting check
I) check if VMX feature is enabled in the BIOS setting. If disabled, the board inspector will show an error message.
II)  check if Hyper-Threading is enabled in the BIOS setting. If enabled, the board inspector will show the warning message.
III) check if VT-d is enabled in the BIOS setting. If disabled, the board inspector will show an error message.

2. Add CPU capability checks.
If a feature is not supported on this processor, the board inspector will show an error message because it would impact ACRN’s ability to function properly.

### CODE REVIEW

**Comment by acrnsi** [comment]
Can one of the admins verify this patch?

**Review by NanlinXie — APPROVED** [review]

File: misc/config_tools/board_inspector/board_inspector.py [hunk 1]
```diff
@@ -112,6 +114,11 @@ def main(board_name, board_xml, args):
                 continue
             module.extract(args, board_etree)
 
+        # Validate the XML against XSD assertions
+        count = validator.validate_board("schema/boardchecks.xsd", board_etree)
+        if count == 0:
+            logging.info("All board checks pass.")
```
Comment (gvancuts) [inline misc/config_tools/board_inspector/board_inspector.py:120]:
```suggestion
            logging.info("All board checks passed.")
```
?

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="every $model in processors/model satisfies exists($model/capability[@id='x2apic'])">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>x2APIC feature is not supported on this processor. ACRN requires this feature to function properly.</xs:documentation>
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="every $model in processors/model satisfies exists($model/capability[@id='popcnt'])">
```
Comment (dbkinder) [inline misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd:126]: How many of these specific feature tests do we need to look for?  Seems like if one of these isn't supported none of them will be, so there could be lots of errors written out (or none).

**Review by dbkinder — CHANGES_REQUESTED** [review]

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+  <xs:assert test="every $model in processors/model satisfies not($model/capability[@id='disable_vmx'])">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Virtual Machine Extensions feature is not enabled.
+ACRN requires this feature to function properly. Please enable it in BIOS.</xs:documentation>
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="every $model in processors/model satisfies exists($model/capability[@id='vmx'])">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Intel(R) Virtualization Technology Extension feature is not supported on this processor. ACRN requires this feature to function properly.</xs:documentation>
```
[the remark this answers is not in the document]
Comment (NanlinXie) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd:29]: We need one more day to implement this, suggest to set low priority and put into backlog and fix in future.

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="every $model in processors/model satisfies exists($model/capability[@id='intel_64'])">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Long mode (x86-64, 64-bit) feature is not supported on this processor. ACRN requires this feature to function properly.</xs:documentation>
```
[the remark this answers is not in the document]
Comment (NanlinXie) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd:35]: Same to above comments, agree?

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="every $model in processors/model satisfies not($model/capability[@id='set_32bit_addr_width'])">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>The width of the physical addresses used for the VMXON region is limited to 32bit. "Intel 64 architecture" feature is not enabled. ACRN requires this feature to function properly.</xs:documentation>
```
[the remark this answers is not in the document]
Comment (NanlinXie) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd:41]: We think there's no BIOS setting to enable this.

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="every $model in processors/model satisfies compare($model/attribute[@id='cpuid_level'], '0x15') &gt;= 0">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Instruction CPUID time stamp counter, nominal core crystal clock information leaf and processor frequency information leaf are not supported. ACRN requires this feature to function properly.</xs:documentation>
```
[the remark this answers is not in the document]
Comment (NanlinXie) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd:47]: Same as above

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+
+  <xs:assert test="every $model in processors/model satisfies exists($model/capability[@id='fast_string'])">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Fast string feature is not enabled. ACRN requires this feature to function properly. Please enable it in BIOS.</xs:documentation>
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="every $model in processors/model satisfies exists($model/capability[@id='erms'])">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Enhanced rep movsb/stosb is not supported on this processor. ACRN requires this feature to function properly.</xs:documentation>
```
[the remark this answers is not in the document]
Comment (NanlinXie) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd:74]: It's unsupported processor model, no BIOS setting could resolve this.

File: misc/config_tools/board_inspector/board_inspector.py (hunk 1 above)
Comment (Kunhui-Li) [reply misc/config_tools/board_inspector/board_inspector.py]: Updated.

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 8]
```diff
@@ -0,0 +1,288 @@
+<?xml version="1.0"?>
+<xs:schema xml:id="root"
+	   xmlns:xs="http://www.w3.org/2001/XMLSchema"
+	   xmlns:acrn="https://projectacrn.org">
+
+  <xs:assert test="contains(//DRHD_INFO[text()], 'DRHD0')">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Intel Virtualization Technology for Directed I/O feature is not enabled.
```
[the remark this answers is not in the document]
Comment (Kunhui-Li) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd]: Updated.

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+ACRN requires this feature to function properly. Please enable it in BIOS.</xs:documentation>
```
[the remark this answers is not in the document]
Comment (Kunhui-Li) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd]: Updated.

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="//processors//core[count(thread) = 1]">
+    <xs:annotation acrn:severity="warning">
+      <xs:documentation>Hyper-Threading (HT) is enabled, while this provides more overall processing power,
```
[the remark this answers is not in the document]
Comment (Kunhui-Li) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd]: Updated.

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+this is also known to adversely impact the realtime behavior.</xs:documentation>
```
[the remark this answers is not in the document]
Comment (Kunhui-Li) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd]: Updated

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+  <xs:assert test="every $model in processors/model satisfies compare($model/attribute[@id='cpuid_level'], '0x15') &gt;= 0">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Instruction CPUID time stamp counter, nominal core crystal clock information leaf and processor frequency information leaf are not supported. ACRN requires this feature to function properly.</xs:documentation>
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="every $model in processors/model
+	  satisfies (($model/attribute[@id='physical_address_bits']) != 0 and ($model/attribute[@id='linear_address_bits']) != 0)">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Can't detect linear/physical address size.</xs:documentation>
```
[the remark this answers is not in the document]
Comment (Kunhui-Li) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd]: It's an unsupported processor model, and the error message has been updated.

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="every $model in processors/model
+	  satisfies not(($model/attribute[@id='physical_address_bits']) > 39 and (not($model/capability[@id='gbyte_pages'])
+	  or not($model/capability[@id='vmx_ept_1gb_page'])))">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Physical-address width over 39 bits must support 1GB large page.</xs:documentation>
```
[the remark this answers is not in the document]
Comment (Kunhui-Li) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd]: It's an unsupported processor model, and the error message has been updated.

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="every $model in processors/model satisfies exists($model/capability[@id='fast_string'])">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Fast string feature is not enabled. ACRN requires this feature to function properly. Please enable it in BIOS.</xs:documentation>
```
[the remark this answers is not in the document]
Comment (Kunhui-Li) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd]: Sorry, the error message has been updated.

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd (hunk 2 above)
Comment (NanlinXie) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd:126]: Yes, it is by design to report every error log we detect, all these error logs will prevent user from booting ACRN.

**Comment by NanlinXie** [comment]
@dbkinder , any more comments?

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd (hunk 2 above)
Comment (dbkinder) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd:126]: OK.

File: misc/config_tools/board_inspector/cpuparser/platformbase.py [hunk 15]
```diff
@@ -135,8 +136,17 @@ def __ne__(self, other):
...
-        r = cls(bits.rdmsr(cpu_id, cls.addr))
+    def rdmsr(cls, cpu_id: int) -> int:
+        try:
+            with open(f'/dev/cpu/{cpu_id}/msr', 'rb') as msr_reader:
+                msr_reader.seek(cls.addr)
+                r = msr_reader.read(8)
+                r = cls(int.from_bytes(r, 'little'))
+        except IOError:
+            logging.critical(f"Missing CPU MSR file at /dev/cpu/{cpu_id}/msr. Check the value of CONFIG_X86_MSR " \
+                             "in the kernel config.  Set it to 'Y' and rebuild the OS. Then rerun the Board Inspector.")
```
[the remark this answers is not in the document]
Comment (Kunhui-Li) [reply misc/config_tools/board_inspector/cpuparser/platformbase.py]: Updated.

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+  <xs:assert test="every $model in processors/model satisfies not($model/capability[@id='disable_vmx'])">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Virtual Machine Extensions feature is not enabled.
+ACRN requires this feature to function properly. Please enable it in BIOS.</xs:documentation>
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="every $model in processors/model satisfies exists($model/capability[@id='vmx'])">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Intel Virtualization Technology Extension (VMX) feature is not supported on this processor. ACRN requires this feature to function properly.</xs:documentation>
```
[the remark this answers is not in the document]
Comment (Kunhui-Li) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd]: Updated.

File: misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+  <xs:assert test="//processors//core[count(thread) = 1]">
+    <xs:annotation acrn:severity="warning">
+      <xs:documentation>Hyper-Threading (HT) is enabled. While this feature can provide more overall processing power,
+hyperthreading can adversely impact predictable real-time performance behavior.</xs:documentation>
+    </xs:annotation>
+  </xs:assert>
+
+  <xs:assert test="every $model in processors/model satisfies not($model/capability[@id='disable_vmx'])">
+    <xs:annotation acrn:severity="error">
+      <xs:documentation>Virtual Machine Extensions feature is not enabled.
```
[the remark this answers is not in the document]
Comment (Kunhui-Li) [reply misc/config_tools/board_inspector/schema/checks/platform_capabilities.xsd]: Updated.

**Review by dbkinder — APPROVED** [review]
Some suggested changes are being deferred to later.

**Pull request merged**